File: 0859-buddy-strings/0859-buddy-strings.cpp
class Solution {
public:
    bool buddyStrings(string s, string goal) {
        if (s.size() != goal.size()) {
            return false;
        }
        string s1;
        string s2;
        set<char> chars;
        bool repeated = false;

        for (size_t i = 0; i < s.size(); ++i) {
            repeated = repeated || !chars.insert(s[i]).second;
            if (s[i] != goal[i]) {
                s1 += s[i];
                s2 += goal[i];
            }
        }
        
        return (s1.size() == 2 && s1 == string(s2.rbegin(), s2.rend())) || (s1.size() == 0 && repeated == true);
    }
};

File: 0859-buddy-strings/0859-buddy-strings.rb
# @param {String} s
# @param {String} goal
# @return {Boolean}
def buddy_strings(s, goal)
  return false if s.size != goal.size
  s1 = String.new
  s2 = String.new
  chars = Set.new
  repeated = false

  s.chars.each_index do |i|
    repeated ||= !chars.add?(s[i]) # Bool
    next if s[i] == goal[i]
    s1 += s[i]
    s2 += goal[i]
  end
  (s1.size == 2 && s1 == s2.reverse) || (s1.size == 0 && repeated == true)
end